File: adapters/node/echo-adapter.mjs
#!/usr/bin/env node

const ambientKeys = [
  "PATH",
  "Path",
  "HOME",
  "USERPROFILE",
  "APPDATA",
  "TEMP",
  "TMP",
  "NODE_OPTIONS",
  "NPM_TOKEN",
  "npm_config_userconfig"
];

let input = "";
process.stdin.setEncoding("utf8");
process.stdin.on("data", (chunk) => {
  input += chunk;
});
process.stdin.on("end", async () => {
  const request = JSON.parse(input);
  const mode = request.mode ?? "success";
  const sleepMs = Number(request.sleepMs ?? 0);

  if (mode === "sleep") {
    await new Promise((resolve) => setTimeout(resolve, Math.max(sleepMs, 0)));
  }

  if (mode === "secret") {
    const marker = "bearer " + "abcdefghijklmnop";
    process.stderr.write(`diagnostic marker ${marker}\n`);
  }

  const payload = {
    runId: request.runId,
    resourceId: request.resource?.id,
    mode,
    message: request.payload?.message ?? "node-adapter-ready",
    envKeys: Object.keys(process.env).sort(),
    ambientEnvNonEmpty: ambientKeys.filter((key) => Boolean(process.env[key])).sort()
  };
  const serialized = JSON.stringify(payload);

  const result = {
    schemaVersion: "agentique.nodeAdapterResult.v1",
    ok: true,
    ready: true,
    outputs: [
      {
        path: "outputs/node-result.json",
        mediaType: "application/json",
        bytes: Buffer.byteLength(serialized, "utf8")
      }
    ],
    artifacts: [
      {
        id: "artifact-node-result-json",
        path: "artifacts/node-result.json",
        viewer: "json",
        redacted: true
      }
    ],
    payload
  };

  process.stdout.write(`${JSON.stringify(result)}\n`);
});

File: adapters/python/echo_adapter.py
#!/usr/bin/env python
import json
import os
import sys
import time


ambient_keys = [
    "PATH",
    "Path",
    "HOME",
    "USERPROFILE",
    "APPDATA",
    "TEMP",
    "TMP",
    "PYTHONPATH",
    "CONDA_PREFIX",
]


def main():
    request = json.load(sys.stdin)
    mode = request.get("mode", "success")
    sleep_ms = int(request.get("sleepMs", 0) or 0)

    if mode == "sleep":
        time.sleep(max(sleep_ms, 0) / 1000)

    if mode == "secret":
        marker = "bearer " + "abcdefghijklmnop"
        print(f"diagnostic marker {marker}", file=sys.stderr)

    if mode == "invalid-json":
        sys.stdout.write("not-json\n")
        return

    if mode == "exit-failure":
        print("intentional adapter failure", file=sys.stderr)
        sys.exit(7)

    payload = {
        "runId": request.get("runId"),
        "resourceId": request.get("resource", {}).get("id"),
        "mode": mode,
        "message": request.get("payload", {}).get("message", "adapter-ready"),
        "envKeys": sorted(os.environ.keys()),
        "ambientEnvNonEmpty": sorted(key for key in ambient_keys if os.environ.get(key)),
    }

    serialized = json.dumps(payload, sort_keys=True)
    result = {
        "schemaVersion": "agentique.pythonAdapterResult.v1",
        "ok": True,
        "ready": True,
        "outputs": [
            {
                "path": "outputs/python-result.json",
                "mediaType": "application/json",
                "bytes": len(serialized.encode("utf-8")),
            }
        ],
        "artifacts": [
            {
                "id": "artifact-python-result-json",
                "path": "artifacts/python-result.json",
                "viewer": "json",
                "redacted": True,
            }
        ],
        "payload": payload,
    }
    json.dump(result, sys.stdout, sort_keys=True)
    sys.stdout.write("\n")


if __name__ == "__main__":
    main()
